Handle timeouts in direct to WooPay checkout flow

ricardo opened the pull request:
Fixes 2482-gh-Automattic/woopay

#### Changes proposed in this Pull Request

This PR adds handling of timeouts and errors in the requests to the direct-to-WooPay checkout flow.

> Covered with tests (or have a good reason not to test in description)

We currently don't cover similar WooPay flows in JS tests. We will most likely continue to cover these flows with E2E tests in the WooPay repository instead.

#### Testing instructions

Testing instructions in Fixes 2513-gh-Automattic/woopay.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

ricardo committed:
Short-circuit WooPay checkout redirect if an error is thrown before retrieving the redirect URL

ricardo committed:
Combine "redirectToWooPaySession" and "redirectToWooPay" methods

ricardo committed:
console.warn thrown errors

ricardo committed:
Add changelog entry

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8214` or branch name `add/timeouts-to-direct-to-woopay-checkout` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/timeouts-to-direct-to-woopay-checkout) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 09e1eb47dda1a102d08e2d154b844849b3e85b59
- Build time: 2024-02-26 16:38:50 UTC

_Note: the build is updated when a new commit is pushed to this PR._

ricardo committed:
Improve comments

ricardo committed:
Rename "woopay_checkout_redirect" to simply "checkout_redirect"

ricardo committed:
Merge branch 'develop' into add/timeouts-to-direct-to-woopay-checkout

ricardo committed:
Improve comments on telemetry TODOs

ricardo commented on client/checkout/woopay/direct-checkout/woopay-direct-checkout.js:
```diff
@@ -127,39 +157,32 @@ class WoopayDirectCheckout {
+					// TODO: Add telemetry as to _how long_ it took to get to this step.
+					window.location.href = woopayRedirectUrl;
+				} catch ( error ) {
+					// TODO: Add telemetry as to _why_ we've short-circuited the WooPay checkout flow.
```
Suggested the creation of an issue for these `TODO` comments in the Epic: https://github.com/Automattic/woopay/issues/2440#issuecomment-1949596649.

cesarcosta99 reviewed (changes requested):
This is tied to https://github.com/Automattic/woopay/pull/2513 so I'm putting it into the same review state. No specific request to this PR just yet.

asumaran committed:
Adjust the selector where href is extracted from

asumaran committed:
Do not break expected behavior if no link is found

During review, we noticed we were redirecting the user to an unrecognized URL when using the Cart Block. Even though the bug is fixed, we're making sure we don't break the merchant's site.

asumaran commented on client/checkout/woopay/direct-checkout/woopay-direct-checkout.js:
```diff
@@ -165,11 +165,17 @@ class WoopayDirectCheckout {
 
+				// If there's no link where to redirect the user, do not break the expected behavior.
+				if ( ! currTargetHref ) {
+					this.teardown();
```
@lovo-h I've added one more commit https://github.com/Automattic/woocommerce-payments/pull/8214/commits/922425b5d4cb5ff995385ca84271d5266efb372f let me know if it makes sense to you.

When @cesarcosta99 found the issue ( https://github.com/Automattic/woopay/pull/2513#pullrequestreview-1889442228 ) I was thinking the issue could happen for various reasons. e.g. The merchant is using a custom a template that uses a different selector. In that case we shouldn't break the normal checkout flow.

Also, I'm putting the `this.teardown();` here not sure if it's OK to do that in case we end the event handler.

lovo-h replied:
@asumaran, that's a good call, and it makes sense 👍

> Also, I'm putting the `this.teardown();` here not sure if it's OK to do that in case we end the event handler.

Since we will be redirected to the Merchant checkout, the class will be automatically torn down. But in this case, it's a good practice to manually remove the listeners by calling the `teardown()` function.

ricardo committed:
Merge branch 'develop' into add/timeouts-to-direct-to-woopay-checkout

cesarcosta99 reviewed (approved):
LGTM!
